core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -419,6 +420,21 @@ public final Boolean readOptionalBoolean() throws IOException {
         return (Map<String, Object>) readGenericValue();
     }
 
+    /**
+     * Read a map of strings to string lists.
+     */
+    public Map<String, List<String>> readMapOfLists() throws IOException {
```

Comment:
I started doing that, but then I realized that it will make it impossible to backport to 2.x without rewriting it back to this (which is still a goal).

I do prefer this approach, particularly for the reader side, and will follow up with a separate PR to do it on master only.